nav2_pure_pursuit_controller/src/pure_pursuit_controller.cpp

```diff
@@ -99,93 +167,262 @@ void PurePursuitController::activate()
+    "Deactivating controller: %s of type pure_pursuit_controller::PurePursuitController\"  %s",
     plugin_name_.c_str());
   global_pub_->on_deactivate();
+  carrot_pub_->on_deactivate();
+  carrot_arc_pub_->on_deactivate();
 }
 
+double PurePursuitController::getLookAheadDistance(const geometry_msgs::msg::Twist & speed)
+{
+  // If using velocity-scaled look ahead distances, find and clamp the dist
+  // Else, use the default look ahead distance
+  double lookahead_dist = 0.0;
+  if (use_velocity_scaled_lookahead_dist_) {
+    lookahead_dist = speed.linear.x * lookahead_time_;
+    lookahead_dist = std::clamp(lookahead_dist, min_lookahead_dist_, max_lookahead_dist_);
+  } else {
+    lookahead_dist = lookahead_dist_;
+  }
+
+  return lookahead_dist;
+}
 
 geometry_msgs::msg::TwistStamped PurePursuitController::computeVelocityCommands(
   const geometry_msgs::msg::PoseStamped & pose,
-  const geometry_msgs::msg::Twist &)
+  const geometry_msgs::msg::Twist & speed)
 {
+  // Transform path to robot base frame
   auto transformed_plan = transformGlobalPlan(pose);
 
-  // Find the first pose which is at a distance greater than the specified lookahed distance
+  // Find look ahead distance and point on path and publish
+  const double lookahead_dist = getLookAheadDistance(speed);
+  auto carrot_pose = getLookAheadMarker(lookahead_dist, transformed_plan);
+  auto carrot_msg = std::make_unique<geometry_msgs::msg::PointStamped>();
+  carrot_msg->header = carrot_pose.header;
+  carrot_msg->point.x = carrot_pose.pose.position.x;
+  carrot_msg->point.y = carrot_pose.pose.position.y;
+  carrot_msg->point.z = 0.01;  // publish right over map to stand out
+  carrot_pub_->publish(std::move(carrot_msg));
+
+  double linear_vel, angular_vel;
+
+  // Find distance^2 to look ahead point (carrot) in robot base frame
+  // This is the chord length of the circle
+  const double carrot_dist2 =
+    (carrot_pose.pose.position.x * carrot_pose.pose.position.x) +
+    (carrot_pose.pose.position.y * carrot_pose.pose.position.y);
+
+  // Find curvature of circle (k = 1 / R)
+  double curvature = 0.0;
+  if (carrot_dist2 > 0.001) {
+    curvature = 2.0 * carrot_pose.pose.position.y / carrot_dist2;
+  }
+
+  // Apply curvature to angular velocity
+  linear_vel = desired_linear_vel_;
+  angular_vel = desired_linear_vel_ * curvature;
+
+  // Make sure we're in compliance with basic constraints
+  applyConstraints(
+    linear_vel, angular_vel,
+    fabs(lookahead_dist - sqrt(carrot_dist2)), lookahead_dist, speed);
+
+  // Collision checking on this velocity heading
+  if (isCollisionImminent(pose, carrot_pose, curvature, linear_vel, angular_vel)) {
+    RCLCPP_ERROR(logger_, "Collision imminent!");
+    throw std::runtime_error("PurePursuitController detected collision ahead!");
+  }
+
+  // populate and return message
+  geometry_msgs::msg::TwistStamped cmd_vel;
+  cmd_vel.header.frame_id = pose.header.frame_id;
+  cmd_vel.header.stamp = clock_->now();
```

Comment:
You could use the stamp of the robot pose here right?

Reply:
Good point

Reply:
Good point, will do